src/openssl.c

```diff
@@ -5487,6 +5488,87 @@ static int xr_setPublicKey(lua_State *L) {
 } /* xr_setPublicKey() */
 
 
+static int xr_setExtensionByNid(lua_State *L, X509_REQ *csr, int target_nid, void* value) {
+	STACK_OF(X509_EXTENSION) *sk = NULL;
+	X509_ATTRIBUTE *attr;
+	int has_attrs=0, idx, *pnid;
+
+	// Replace existing if it's there. Extensions are stored in a CSR in an interesting way:
+	//
+	// They are stored as a list under either (most likely) the "official"
+	// NID_ext_req or under NID_ms_ext_req which means everything is stored
+	// under a list in a single "attribute" so we can't use X509_REQ_add1_attr
+	// or similar.
+	//
+	// Instead we have to get the extensions, find and replace the SAN if it's
+	// in there, then *replace* the extensions in the list of attributes. (If
+	// we just try to add it the old ones are found first and don't take
+	// priority)
+
+	has_attrs = X509_REQ_get_attr_count(csr);
+	sk = X509_REQ_get_extensions(csr);
+
+	if (!X509V3_add1_i2d(&sk, target_nid, value, 0, X509V3_ADD_REPLACE)) {
+		goto error;
+	}
+
+	// Delete the old extensions attribute, and replace with the our new ones
```

Comment:
Probably, yes. I'll try that.

If the delete fails then we could end up with two extensions attrs (causing the second to be ignored most likely) but that is likely a better failure mode.